### PR TITLE
Initial Android cross-compiler

### PR DESCRIPTION
**Pull request by directhex**
- [x] Only works on Mac due to `--host=` value (we must pass a `--host` triplet when we pass `--target`, and need to infer the correct one to use). This will cause CI failures until fixed, as we only build the runtime on Linux in CI.
- [x] Audit the configure flags, I sorta guessed based on gluing together the AOT flags from iOS/WASM and the JIT flags from Android
- [x] Less brute-force approach to not using the same CFLAGS etc for "real" runtime as cross-compiler. This was a fast way to filter out things I didn't want without breaking iOS etc, but it's not elegant and will cause problems in the future
- [x] Move AOT building to a new subset, so we can do AOT-only builds (i.e. avoid doing full builds on Mac/Windows)
- [ ] Add Windows version (i.e. Windows builds with vcxproj, we need a cross-compiler vcxproj)
- [x] Add Yaml for all the runtime builds (maybe a new yaml, rather than runtime.yml?)

### CODE REVIEW

File: src/mono/mono.proj [hunk 1]
```diff
@@ -750,7 +772,19 @@
...
       <_MonoAotCrossCCLDFLAGSOption Condition="@(_MonoAotCrossCCLDFLAGS->Count()) &gt; 0">CCLDFLAGS="@(_MonoAotCrossCCLDFLAGS, ' ')"</_MonoAotCrossCCLDFLAGSOption>
 
-      <_MonoAotCrossConfigureCommand>$(MonoProjectRoot)configure @(_MonoAotCrossConfigureParams, ' ') @(_MonoAotCrossAC_VARS, ' ') $(_MonoAotCrossCFLAGSOption) $(_MonoAotCrossCXXFLAGSOption) $(_MonoAotCrossCPPFLAGSOption) $(_MonoAotCrossCXXCPPFLAGSOption) $(_MonoAotCrossLDFLAGSOption) $(_MonoAotCrossCCLDFLAGSOption) $(_MonoCCOption) $(_MonoCXXOption) $(_MonoAROption) $(_MonoASOption) $(_MonoCPPOption) $(_MonoCXXCPPOption) $(_MonoDLLTOOLOption) $(_MonoLDOption) $(_MonoOBJDUMPOption) $(_MonoRANLIBOption) $(_MonoCMAKEOption) $(_MonoSTRIPOption)</_MonoAotCrossConfigureCommand>
+      <_MonoAotCrossCCOption Condition="'$(TargetsAndroid)' != 'true'">$(_MonoCCOption)</_MonoAotCrossCCOption>
+      <_MonoAotCrossCXXOption Condition="'$(TargetsAndroid)' != 'true'">$(_MonoCXXOption)</_MonoAotCrossCXXOption>
+      <_MonoAotCrossAROption Condition="'$(TargetsAndroid)' != 'true'">$(_MonoAROption)</_MonoAotCrossAROption>
```
Comment (vargaz) [inline src/mono/mono.proj:777]: We don't have most of these options in the cmake build.

File: src/mono/netcore/sample/Android/Makefile [hunk 2]
```diff
@@ -12,7 +12,9 @@ run:
 	/p:TargetArchitecture=$(MONO_ARCH) \
 	/p:TargetOS=Android \
 	/p:Configuration=$(MONO_CONFIG) \
-	/p:DeployAndRun=true
+	/p:DeployAndRun=true \
```
Comment (vargaz) [inline src/mono/netcore/sample/Android/Makefile:15]: The samples should not use aot imho.

**Comment by vargaz** [comment]
Comments:
- the -O=-float32 argument to the aot compiler should be removed
- -O=-simd should be added to work around a ton of SIMD problems
- it would be nice to print out the assembly paths for the assemblies which fail aot, its hard to find out which ones because of the parallelism.

**Comment by steveisok** [comment]
@directhex Let's try to land the android x64 host + all android targets in a separate PR.  If OSX comes along cheaply, I'm fine with including that as well.  The rest I'll list in an issue and we can chip away w/ individual PR's.

**Comment by directhex** [comment]
I expect this to fail right now on not-linux_x64. Need to figure out how to set android_ndk_root for mac (it's in our images already but not exported as an env var)

File: src/mono/mono.proj [hunk 3]
```diff
@@ -457,7 +471,7 @@
         <Destination>$(RuntimeBinDir)$(MonoStaticFileName)</Destination>
       </_MonoRuntimeArtifacts>
       <_MonoRuntimeArtifacts Include="$(_MonoAotCrossFilePath)">
-        <Destination>$(RuntimeBinDir)cross\mono-aot-cross</Destination>
+        <Destination>$(RuntimeBinDir)cross\$(TargetOS)-$(TargetArchitecture)\mono-aot-cross</Destination>
       </_MonoRuntimeArtifacts>
```
Comment (vargaz) [inline src/mono/mono.proj:475]: Why is this needed ? Isn't the runtime pack already os/arch specific ?

File: src/mono/mono.proj [hunk 4]
```diff
@@ -466,10 +480,10 @@
         <Destination>$(RuntimeBinDir)\opt</Destination>
       </_MonoRuntimeArtifacts>
       <_MonoRuntimeArtifacts Condition="'$(MonoAOTBundleLLVMOptimizer)' == 'true'" Include="$(MonoAOTLLVMDir)\bin\llc">
-        <Destination>$(RuntimeBinDir)cross\llc</Destination>
+        <Destination>$(RuntimeBinDir)cross\$(TargetOS)-$(TargetArchitecture)\llc</Destination>
       </_MonoRuntimeArtifacts>
```
Comment (vargaz) [inline src/mono/mono.proj:484]: llc/opt are not target arch specific.

File: src/mono/monoaotcross.proj [hunk 5]
```diff
@@ -0,0 +1,40 @@
+<Project Sdk="Microsoft.Build.Traversal" DefaultTargets="BuildMonoCrossAllTargets">
```
Comment (vargaz) [inline src/mono/monoaotcross.proj:1]: Why is this needed ? Why can't the cross compilers be built normally as on ios ?

Comment (directhex) [reply src/mono/monoaotcross.proj:1]: iOS assumes a 1:1 mapping between "host OS/arch" and "target OS/arch". "Build the iOS ARM64 cross compiler" _always_ means "Build the macOS x64 to iOS ARM64 cross compiler". The build pipelines and directory structures are built around this 1:1 assumption.

Right now, this would mean we can only build _one_ Android x64 cross compiler - Linux x64 (the build platform) to Android x64.

We need to decouple the entire 1:1 assumption in order to be able to build multiple hosts for the same target.

File: src/mono/mono.proj (hunk 3 above)
Comment (directhex) [reply src/mono/mono.proj:475]: We need a clean way to have multiple cross compilers next to each other, in order to enable a user to build an apk which supports multiple target architectures (which is common on Android).

File: src/mono/mono.proj (hunk 4 above)
Comment (directhex) [reply src/mono/mono.proj:484]: Assuming that this isn't true is easier for now, i.e. put "the `llc/opt` needed for this `mono-aot-cross` in the same bundle" rather than dependency tracking & public artifact publishing of the llvm binaries for end user consumption. We can clean that story up later.

**Comment by directhex** [comment]
Working:
- Linux x64
- OSX x64

Failing:
- Windows x64
- Windows x86

Building, but building wrong:
- Linux arm64

**Comment by directhex** [comment]
Punting everything except Mac x64 and Linux x64 host platform builds, to speed things along.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.sfxproj [hunk 6]
```diff
@@ -0,0 +1,54 @@
+<Project>
+  <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />
+  <Import Project="Sdk.props" Sdk="Microsoft.DotNet.SharedFramework.Sdk" />
+
+  <PropertyGroup>
+    <SkipValidatePackage>true</SkipValidatePackage>
+    <PlatformPackageType>RuntimePack</PlatformPackageType>
+    <SharedFrameworkName>Microsoft.NETCore.App.MonoCrossAOT</SharedFrameworkName>
+    <OverridePackageId>Microsoft.NETCore.App.MonoCrossAOT.$(RuntimeIdentifier).targeting.$(TargetCrossRid)</OverridePackageId>
```
Comment (steveisok) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.sfxproj:9]: This should be `Microsoft.NetCore.App.Runtime.AOT.{host-rid}.Cross.{target-rid}.{version}`

**Review by steveisok — APPROVED** [review]
Looks good to me

**Comment by ViktorHofer** [comment]
181 commits 🤨?

**Comment by directhex** [comment]
It was originally based on a long term feature branch not master, that can confuse the GH UI

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.sfxproj [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  </Target>
+
+  <Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />
+  <Import Project="Sdk.targets" Sdk="Microsoft.DotNet.SharedFramework.Sdk" />
+
+  <Target Name="ValidateProperties" BeforeTargets="GenerateNuspec">
+    <Error Condition="'$(TargetCrossRid)' == ''" Text="TargetCrossRid not set" />
+  </Target>
+
+  <Target Name="GetFilesToPublish">
```
Comment (jkoritzinsky) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.sfxproj:40]: What was the need for adding this target? At first glance it looks like it matches the one in the SDK.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/monocrossaot.sfxproj [hunk 8]
```diff
@@ -0,0 +1,27 @@
+<Project Sdk="Microsoft.Build.Traversal" DefaultTargets="BuildMonoCrossAllTargets">
```
Comment (jkoritzinsky) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/monocrossaot.sfxproj:1]: If we’re using the traversal SDK, then we should hook into the Traversal targets instead of defining our own.

**Comment by directhex** [comment]
@jkoritzinsky not really using the traversal SDK, just a leftover from a prior incarnation of that XML, which didn't _break_ whilst not being correct either.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/monocrossaot.sfxproj [hunk 9]
```diff
@@ -0,0 +1,29 @@
+<Project Sdk="Microsoft.NET.Sdk" DefaultTargets="BuildMonoCrossAllTargets">
+  <Sdk Name="Microsoft.DotNet.SharedFramework.Sdk" />
```
Comment (jkoritzinsky) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/monocrossaot.sfxproj:2]: If you're not going to use any of the built-in targets, then I'd suggest just using the NoTargets SDK.

```suggestion
<Project Sdk="Microsoft.Build.NoTargets" DefaultTargets="BuildMonoCrossAllTargets">
```

File: src/mono/monoaotcross.proj [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
```
Comment (jkoritzinsky) [inline src/mono/monoaotcross.proj:1]: Same here with the SDK.

```suggestion
<Project Sdk="Microsoft.Build.NoTargets" DefaultTargets="BuildMonoCrossAllTargets">
```

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.sfxproj [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+  <ItemGroup>
+    <NativeRuntimeAsset Include="$(MonoCrossAOTDir)$(TargetCrossRid)\**" TargetPath="tools/" />
+  </ItemGroup>
+
+  <Target Name="AddRuntimeFilesToPackage" DependsOnTargets="ResolveRuntimeFilesFromLocalBuild">
+  </Target>
+
+  <Target Name="AddFrameworkFilesToPackage" DependsOnTargets="ResolveLibrariesFromLocalBuild">
+  </Target>
```
Comment (jkoritzinsky) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.sfxproj:31]: If you aren't putting anything in these targets then you can remove them and remove setting `GetSharedFrameworkFilesForReadyToRunDependsOn`.

In addition, if all the files are native, then there are more optimizations we can make here to reduce the work we do.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.sfxproj [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+  <Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />
+  <Import Project="Sdk.targets" Sdk="Microsoft.DotNet.SharedFramework.Sdk" />
+
+  <Target Name="ValidateProperties" BeforeTargets="GenerateNuspec">
+    <Error Condition="'$(TargetCrossRid)' == ''" Text="TargetCrossRid not set" />
+  </Target>
```
Comment (jkoritzinsky) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.sfxproj:38]: You should be able to put this before the `<Import>` tags above and switch the `<Import>` tags back to the `Sdk="Microsoft.NET.Sdk"` attribute and the `<Sdk Name="Microsoft.Dotnet.SharedFramework.Sdk" />` tag.

**Review by jkoritzinsky — APPROVED** [review]

**Comment by jkoritzinsky** [comment]
Just a few cleanup suggestions, but other than those LGTM.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.sfxproj (hunk 12 above)
Comment (directhex) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.sfxproj]: Can you check the fixes for this one (e157f3f72c5) are correct?

**Comment by directhex** [comment]
@jkoritzinsky I've switched monoaotcross.proj back to Traversal. With NoTargets, the following error happens during a clean-tree build:

```
                     /Users/directhex/Projects/runtime/.dotnet/sdk/5.0.100/NuGet.RestoreEx.targets(19,5): error : Invalid framework identifier ''. [/Users/directhex/Projects/runtime/Build.proj]
```

**Comment by jkoritzinsky** [comment]
I'm that case I'd suggest hooking the defined targets in the project into the targets in the Traversal SDK, but that's not required.

**Comment by ViktorHofer** [comment]
> With NoTargets, the following error happens during a clean-tree build:

The NoTargets SDK needs a TargetFramework specified. If you want to stick with NoTargets you can fix it by adding this code block:
```
<PropertyGroup>
    <TargetFramework>$(NetCoreAppToolCurrent)</TargetFramework>
</PropertyGroup>
```

**Comment by directhex** [comment]
Thanks @ViktorHofer

**Comment by steveisok** [comment]
The windows test failure will be resolved by https://github.com/dotnet/runtime/pull/46828

File: eng/pipelines/mono/templates/build-job.yml [hunk 13]
```diff
@@ -13,6 +13,9 @@ parameters:
   isOfficialBuild: false
   crossBuild: false
   crossrootfsDir: ''
+  dependOnEvaluatePaths: false
```
Comment (safern) [inline eng/pipelines/mono/templates/build-job.yml:16]: This is outdated, I updated this to: `dependOnCheckoutBundle` here: https://github.com/dotnet/runtime/commit/604660b37a0d416155a97cbd09c1edc101a1a1cb

File: src/installer/pkg/sfx/Microsoft.NETCore.App/monocrossaot.sfxproj [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+  </PropertyGroup>
+
+  <ItemGroup>
+    <MonoAotTargetRids Include="$(MonoAotTargets.Split(';'))" />
+  </ItemGroup>
+
+  <Target Name="RestoreSubProject">
+    <MSBuild Targets="Restore"
+             Projects="$(MSBuildThisFileDirectory)Microsoft.NETCore.App.MonoCrossAOT.sfxproj" />
+  </Target>
```
Comment (ViktorHofer) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/monocrossaot.sfxproj:14]: This causes a separate restore invocation to be triggered which we don't want for different reasons (reliability, performance, etc.). Why is this necessary? If you use the Traversal SDK, you can just add ProjectReference to that project if you don't want to expose it in Subsets.props.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/monocrossaot.sfxproj [hunk 15]
```diff
@@ -0,0 +1,28 @@
+<Project Sdk="Microsoft.Build.NoTargets" DefaultTargets="BuildMonoCrossAllTargets">
```
Comment (ViktorHofer) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/monocrossaot.sfxproj:1]: You don't need a default target here and just hook that target as `AfterTargets="Build"` to run after the build which is the default target.

Comment (ViktorHofer) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/monocrossaot.sfxproj:1]: @jkoritzinsky just to make sure, the sfxproj SDK doesn't change the default target, right?

File: src/mono/monoaotcross.proj [hunk 16]
```diff
@@ -0,0 +1,46 @@
+<Project Sdk="Microsoft.Build.NoTargets" DefaultTargets="BuildMonoCrossAllTargets">
```
Comment (ViktorHofer) [inline src/mono/monoaotcross.proj:1]: Same comment about the default target. Just remove the attribute here and use an aftertargets="Build" in the `BuildMonoCrossAllTargets` target.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/monocrossaot.sfxproj (hunk 15 above)
Comment (jkoritzinsky) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/monocrossaot.sfxproj:1]: You are correct. The SharedFX sdk still uses the Build target as the entry point.

File: src/mono/netcore/sample/Android/Makefile [hunk 17]
```diff
@@ -1,6 +1,8 @@
 MONO_CONFIG=Release
 MONO_ARCH=x64
 DOTNET := ../../../../.././dotnet.sh
+USE_LLVM=true
+FORCE_AOT=false
```
Comment (vargaz) [inline src/mono/netcore/sample/Android/Makefile:6]: Other samples name this variable 'AOT'.

File: src/mono/netcore/sample/iOS/Makefile [hunk 18]
```diff
@@ -1,7 +1,8 @@
 MONO_CONFIG=Debug
 MONO_ARCH=x64
 DOTNET := ../../../../.././dotnet.sh
-USE_LLVM=True
+USE_LLVM=true
+FORCE_AOT=false
```
Comment (vargaz) [inline src/mono/netcore/sample/iOS/Makefile:6]: Ditto.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/monocrossaot.sfxproj (hunk 14 above)
Comment (directhex) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/monocrossaot.sfxproj:14]: The build was failing without the explicit second call to restore. I know it's ugly, but it's the workaround I found which worked.

Your suggestion is adding a ProjectReference to Microsoft.NETCore.App.MonoCrossAOT.sfxproj in monocrossaot.sfxproj, and switching Microsoft.NETCore.App.MonoCrossAOT.sfxproj to Traversal instead of NoTargets?

Comment (ViktorHofer) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/monocrossaot.sfxproj:14]: > Your suggestion is adding a ProjectReference to Microsoft.NETCore.App.MonoCrossAOT.sfxproj in monocrossaot.sfxproj, and switching Microsoft.NETCore.App.MonoCrossAOT.sfxproj to Traversal instead of NoTargets?

Right, but only if that project isn't restored yet. What was the failure that you were hitting?

File: src/mono/mono.proj [hunk 19]
```diff
@@ -417,18 +431,21 @@
       <_MonoAotCMakeBuildCommand>cmake --build . --target install --config $(Configuration)</_MonoAotCMakeBuildCommand>
       <_MonoAotCMakeBuildCommand Condition="'$(MonoVerboseBuild)' == 'true'">$(_MonoAotCMakeBuildCommand) --verbose</_MonoAotCMakeBuildCommand>
       <_MonoAotCMakeBuildCommand Condition="'$(_MonoUseNinja)' != 'true'">$(_MonoAotCMakeBuildCommand) --parallel $([System.Environment]::ProcessorCount)</_MonoAotCMakeBuildCommand>
+      <_MonoAotPrebuiltOffsetsFile>$(ArtifactsObjDir)\monooffsetfiles\$(PlatformConfigPathPart)\cross\$([System.IO.Path]::GetFileName('$(MonoAotOffsetsFile)'))</_MonoAotPrebuiltOffsetsFile>
     </PropertyGroup>
```
[the remark this answers is not in the document]
Comment (directhex) [reply src/mono/mono.proj:435]: It can be. Originally it was a little easier for local testing workflow if rm -r artifacts/bin/mono/ didn't touch my local copy of the offset files

**Comment by directhex** [comment]
Green with correct nupkgs on internal https://dev.azure.com/dnceng/internal/_build/results?buildId=950402&view=results

**Comment by directhex** [comment]
None of the remaining tests are related to stuff I've touched (and the one failure isn't my fault).

Anything more to add, @vargaz @safern @ViktorHofer @akoeplinger ?

File: eng/pipelines/mono/templates/build-job.yml [hunk 20]
```diff
@@ -101,6 +112,13 @@ jobs:
       - script: $(Build.SourcesDirectory)\eng\common\init-tools-native.cmd -InstallDirectory $(Build.SourcesDirectory)\native-tools -Force
         displayName: Install native dependencies
 
+    - ${{ if eq(parameters.monoCrossAOTTargetOS, 'Android') }}:
+      - task: DownloadPipelineArtifact@2
+        displayName: Download AOT offset files
+        inputs:
+          artifact: Mono_Offsets_Android
+          path: '$(Build.SourcesDirectory)/artifacts/obj/mono/offsetfiles'
```
Comment (akoeplinger) [inline eng/pipelines/mono/templates/build-job.yml:120]: do we still need this here if it's already in global-build-job.yml?

File: eng/pipelines/runtime-official.yml [hunk 21]
```diff
@@ -122,6 +122,37 @@ stages:
...
+  - template: /eng/pipelines/common/platform-matrix.yml
+    parameters:
+      jobTemplate: /eng/pipelines/common/global-build-job.yml
+      runtimeFlavor: mono
+      buildConfig: release
+      platforms:
+      - OSX_x64
+      - Linux_x64
+      jobParameters:
+        buildArgs: -s mono+libs+host+packs -c $(_BuildConfig)
```
Comment (akoeplinger) [inline eng/pipelines/runtime-official.yml:146]: do we really need the `libs` subset here?

File: eng/pipelines/runtime-official.yml [hunk 22]
```diff
@@ -122,6 +122,37 @@ stages:
...
           name: MonoRuntimePacks
 
+  # Build Mono AOT offset headers once, for consumption elsewhere
+  #
+  - template: /eng/pipelines/common/platform-matrix.yml
+    parameters:
+      jobTemplate: /eng/pipelines/mono/templates/generate-offsets.yml
+      buildConfig: release
+      platforms:
+      - Android_x64
```
Comment (akoeplinger) [inline eng/pipelines/runtime-official.yml:132]: maybe add a comment here that this will actually generate offsets for all platforms, not just x64.

File: eng/pipelines/runtime.yml [hunk 23]
```diff
@@ -183,6 +183,22 @@ jobs:
...
 
+# Build Mono AOT offset headers once, for consumption elsewhere
+# Only when mono changed
+#
+- template: /eng/pipelines/common/platform-matrix.yml
+  parameters:
+    jobTemplate: /eng/pipelines/mono/templates/generate-offsets.yml
+    buildConfig: release
+    platforms:
+    - Android_x64
```
Comment (akoeplinger) [inline eng/pipelines/runtime.yml:194]: same comment here

File: eng/pipelines/runtime.yml [hunk 24]
```diff
@@ -459,6 +475,34 @@ jobs:
           eq(dependencies.checkout.outputs['SetPathVars_mono.containsChange'], true),
           eq(variables['isFullMatrix'], true))
 
+#
+# Build Mono release Android AOT cross-compiler
+# Only when mono changed
```
Comment (akoeplinger) [inline eng/pipelines/runtime.yml:480]:
```suggestion
# Only when mono or installer changed
```

File: src/mono/mono.proj [hunk 25]
```diff
@@ -378,8 +380,20 @@
...
+      <MonoAotAbi Condition="'$(Platform)' == 'arm64'">aarch64-v8a-linux-android</MonoAotAbi>
+      <MonoAotAbi Condition="'$(Platform)' == 'arm'">armv7-none-linux-androideabi</MonoAotAbi>
+      <MonoAotAbi Condition="'$(Platform)' == 'x86'">i686-none-linux-android</MonoAotAbi>
+      <MonoAotAbi Condition="'$(Platform)' == 'x64'">x86_64-none-linux-android</MonoAotAbi>
+      <MonoAotOffsetsFile>$(MonoObjDir)cross/offsets-$(Platform)-android.h</MonoAotOffsetsFile>
+    </PropertyGroup>
+
     <PropertyGroup>
-      <MonoLibClang Condition="'$(MonoLibClang)' == ''">$(XcodeDir)/Toolchains/XcodeDefault.xctoolchain/usr/lib/libclang.dylib</MonoLibClang>
+      <MonoLibClang Condition="$([MSBuild]::IsOSPlatform('OSX')) and '$(MonoLibClang)' == ''">$(XcodeDir)/Toolchains/XcodeDefault.xctoolchain/usr/lib/libclang.dylib</MonoLibClang>
```
Comment (akoeplinger) [inline src/mono/mono.proj:395]: we should probably use the libclang.dylib from the ANDROID_NDK_ROOT here too instead of the one from xcode.

**Review by akoeplinger — APPROVED** [review]
Looks great, left a couple comments but nothing that should hold off merging this :)
